Add glossary and components page. (#12)
pymeasure/experiment_control_protocol: components.md

# Components

This page provides details on the main component/entity types that make up a deployment of the ECP.

## Director
A Director manages a part of or the entire setup, orchestrating the Components according to the needs of the experiment.
It can, among other things, 
* issue commands to Actors
* request data from individual Actors
* read data from an Observer

Potentially a GUI could be attached here too.

## Actor

> A Actor is a component that interfaces with a (hardware) Device and that has a specific API on the ECP side.

Reviewer: ```suggestion
An Actor is a component that interfaces with a (hardware) Device and that has a specific API on the ECP side.
```

> A Actor must contain a Driver object dealing with communication with the Device.

Reviewer: ```suggestion
An Actor must contain a separate Driver object which communicates with the Device.
```

Thoughts?
:::

## Driver
A Driver is a component that interfaces with a (hardware) Device in a manner we don't specify, and that has a specific API on the ECP side.
A driver must contain some object dealing with hardware communication.
This may be a `pymeasure.Instrument` instance or something from another instrument library.
This is the place where all instrument libraries (including pymeasure) wire their hardware interface classes into ECP.

Concerning the ECP, we draw the abstraction boundary at the Driver -- the details on how this communicates with a Device (SCPI, dll, ...) should not be relevant for the protocol details.
What we define is how the other ECP components interact with the Driver, how it determines and announces its capabilities, etc.

A driver must contain/manage
* An object that communicates with the connected Device, including managing its lifecycle (init, operations, shutdown)
* The name of the connected instrument
* A list of available Parameters (properties/attributes to get/set) and Actions(methods to call)
* `set`/`get`/`get_all`/`call` interfaces (for incoming commands to use to act on Parameters and Actions)

> 

Reviewer: "I see now what you did there" - so I....well. What is described here now as Driver, is what I had in mind as Actor, with the Actor being allowed to do a bit more stuff. I guess it may make sense to separate the two further, where it then also starts making more sense to rename the Actor to _Processor_. For me the Driver actually was the `pymeasure.Instrument`, so not a newly defined/specified component of the ECP in itself, but whatever hardware communication object that may exist (already, or at least separate from the ECP). 

In that sense, I think we could add a _Processor_ component which is then more what is described above as Actor, while the Driver here is renamed as Actor, while the "some object dealing with hardware communication" remains becomes the Driver, outside of the ECP. However, as a different thing, we could leave it as bilderbuchi now suggested, and rename the Actor to Processor, without giving the "object dealing with hardware communication" a separate name for our discussion (or at least choose a different one to "Driver"). 

In general, how it is described here, sounds to me, when going bottom-up and ending at the Driver, exactly as yaq, the Driver being a yaq daemon. In yaq (or at least in the python examples I saw, I did not look at all implementations, in both in python and otherwise), the Driver communicates directly with the Device, instantiating e.g. a serial connection, having methods defined which deal exclusively with hwardware (HW) -comms and HW-control, and exposing those to clients. What was important to me, was to separate this ECP communication layer from the HW communication layer, which might be e.g. a pymeasure.Instrument. This would clearly separate what `pymeasure` (and many homebrew hardware drivers I have seen) does from what ECP should do - however, I realise that this might "not be up to the ECP" to decide, whether people who implement an ECP Driver (in the terms described in the currently proposed version) do "play nice" and separate it, or not. 
If the two are separated, and things stay that much modular, it becomes easy to re-use these software parts in other frameworks, in case that is wanted/necessary. We might want to encourage possible adopters to stay with this nice new thing we defined/built, and trying to lock them in is not the way I would like to go about it. Most likely they will see it in advance and not start at all - we should convince possible adopters with the content itself, so that they stay because they like it, not because it is a hassle to move their definitions to a different type of framework.

Author: Thank you for your thoughts, I have to reply later.

Author: I'm not sure where you seem to(?) get the idea from that I want to "lock in" people into this framework somehow -- that is not my intention at all! I'd like to fix whatever gave that impression.

On the contrary, I'm aiming for a design where you can use those other objects _unchanged_, and all you have to maintain is an interface class (the `Driver` subclass) that wires things correctly between a `Driver.list_parameters()` and however `self.interface` yields that information, or what `Driver.get_param(name)` needs to do to get this from `self.interface` (in case of pymeasure.Instrument, this might a simple `return self.interface.getattr(name)`.

Author: > What was important to me, was to separate this ECP communication layer from the HW communication layer, which might be e.g. a pymeasure.Instrument. 

Does what I sketched in my comments fulfill this? At some point, you will need interface between ECP and HW, and I intended (misunderstanding the intention behind the original discussion comments -- sorry!) the Driver to be that place:
* ECP messages going in/out
* internally mapping the messages to what the contained HW communication layer/object understands
* the HW communication layer responds, that reply gets transformed into an appropriate message

Reviewer 2: We definitely should talk about that part (driver actor, hardware...) in an issue, where the discussion is more easy to find.

I thought as well, that a Driver is a pymeasure instrument (or similar) and the actor is the interface from the network to the driver.

Reviewer 2: I like it. Can be resolved from my side

Author: I re-read @bklebel's initial comment here, and I think it's resolved the way he wanted to see.
The only point I'd still like to clarify is where he perceived "lock in" potential -- I want very much to avoid that!

Reviewer: Thanks for all your work, I was indisposed the last weeks, sorry for the late reply. @bilderbuchi is correct, how you changed it now is actually quite what I had in mind, so this is resolved for me. The "lock in" potential I saw, was if Actors and Drivers are not separated. Because, if I have something like that (typical in the yaq daemons I saw, and a the Tango python code I looked at), it is quite difficult to take only the Driver part, and use it somewhere else (i.e. if I find a Tango device server for a specific instrument, it does not help me at all, as long as I do not set up the whole Tango thing in my setup). 
If I search for a framework to work with, and I find one which tells me to first write a proper Driver class, and then this will be taken on a piggyback-ride by an Actor, I personally feel safer about the time I might invest into the Driver-development, because if I discovered that I don't like how the Actor carries my Driver around, I can just pluck the Driver out, and put it into something else. If Driver and Actor are fused together as a standard, this makes the effort to separate the two typically so big that it's quicker to just grab the Device's manual and write a new Driver from scratch, i.e. this produces a high threshold to change. 
How it is written now is fine with me.

Author: Ah I see what you mean! Since the whole thing is meant to interface/consume other libraries, I don't see much of a danger of that happening.

    - If caching is included, the `get*` interfaces must include a configurable cache timeout and a way to force fetching a fresh value. 
* Concurrent access management/locking
* Logging configuration

:::{admonition} TODO
We might want to add the notion of "Channels", especially for the multi-Director stuff
:::

## Parameter
A Parameter is a property (in the English, not the Pythonic sense) of the device represented by a Driver
It has a name and can be read(`get`) or `set`.

:::{note}
Recent values may be cached in the Driver.
:::

It may correspond closely to _attributes_ or Python (or PyMeasure) _properties_ of the instrument interface classes.
It may have unit information, that is used when sending data over the network.

## Procedures
Sequences of steps that make up experiment runs, e.g. PyMeasure procedures.
These instructions could be consumed by a Director and trigger a sequence of commands (ramps, loops, conditionals,...).

:::{admonition} TODO
This is a placeholder, we have not fleshed out the concept yet
:::

## Coordinator
A component primarily concerned with routing/coordinating the message flow between other components.

> It represent the intermediate zmq brokers, proxies or somesuch.

Reviewer: ```suggestion
It represents the intermediate zmq brokers, proxies or somesuch.
```

Multiple coordinator instances may be necessary for large deployments, but a single coordinator instance should be sufficient for operation.

The presence of a coordinator should avoid the complexity/scaling of a purely point-to-point approach. 

> Currently, this corresponds to the zmq majordomo broker, and/or a zmq proxy.

Reviewer: ```suggestion
Currently, this corresponds to the something like a zmq majordomo broker, and/or a zmq proxy.
```

## Observer
A component that receives data from other components, e.g. for logging, storage, or plotting, either directly in a streaming fashion, batched, or delayed.
It only consumes message streams, but does not command `actors`.

:::{note} Depending on setup, some commanding might be necessary, e.g. to subscribe/register.
:::

## Notes 

> ### Actor Driver separation

Reviewer: I would like to update this section once we have agreed on what actually Drivers, Actors, and maybe Processors should do. Currently, it does not really represent what I wanted to say when I wrote it, especially regarding how the low-entry-threshold would be achieved, and a potential traversal of building a system might happen, at least with the current terminology.

Author: Sure, maybe it will be folded into the different sections and the rest discarded.

Author: This became streamlined into a new section "Complexity scaling", maybe this expresses your intention better now, and we can iterate/improve on that?

# Components

This page provides details on the main components/entities of the ECP.

## Director

> The Director is the main controller orchestrating the whole thing, an object which can 

Reviewer 2: I'd make it more general, that you can have more than one director:
```suggestion
A Director manages a part or the entire setup orchestrating the actors and observers according to the needs of the experiment.
```

Author: I see the use case for having more than one coordinator,  but I imagined that the director is a singleton. Otherwise, we don't have a single source of truth for commands.
If you have two directors, and one tells an instrument to put Voltage to 100, and another to 5, what then?

Reviewer 2: > If you have two directors, and one tells an instrument to put Voltage to 100, and another to 5, what then?

Then the instrument puts the voltage to 100 and with the next command to 5.

## My reasons for not only one director.
- We need regularly more than one director (two or more people working independently on different parts of the beamline).
- Sometimes we need access to the same VISA resource (stepper motor card controlling 6 motors), which is difficult, as only one program may open that resource. The separation of director and actor will allow to use the same card from different directors.
- I envison the director to be (potentially) a simple script (create director communication part, send commands to actors, read actors, loop the last two steps a few times, print results, done). Afterwards you use another script as director. Forcing a single director seems more difficult

## Possible solution
- We could allow to lock/reserve actors for a director, such that an actor refuses commands (all or just potentially harmful commands like setters) from other directors than the original one.

If we do not reach an agreement within a few comments, we should move it to its own issue, I'd say.

Author: > more people working _independently_ on _different_ parts 

OK, this i can understand, but not two directors "fighting" over control of one voltage (say). 
You raise a good point about a multichannel device with different channels for different users, that was also what I thought of later. 
Locking would again make that usecase impossible, though. We could lock individual Channels, but that would not make sense for all instruments.

Let's not restrict to one now, and I'll open an issue about how to deal with control contention.

* issue commands to actors
* possibly request data from individual actors
* possibly read data from an observer

Potentially a GUI could be attached here too.

## Actor
The Actor is a component which (typically) instantiates one or more Driver class and/or does some kind of processing.
Actions which the actor can take could be anything from measuring on a regular basis ("acting" as a sensor), commanding hardware devices through the use of the driver to do X or Y, processing some data and re-emitting results (e.g. a PID controller).

Potentially a GUI could be attached here too.

## Driver

> A Driver is a component that interfaces with some hardware in a manner we don't specify, and that has a specific API on the ECP side.

Author: I envision all drivers to have a homogeonous interface, and do some bookkeeping about the connected instrument. This means that the interaction with Actors will be easier to specify, and all differences between instruments or instrument libraries are hidden/abstracted away behind the Driver.

I'll flesh this out some with a proposal.

> A Driver is a component that interfaces with some hardware in a manner we don't specify, and that has a specific API on the ECP side.

Author: I extended the description, please continue feedback in the new text as appropriate.

# Components

This page provides details on the main component/entity types that make up a deployment of the ECP.

## Director
A Director manages a part of or the entire setup, orchestrating the actors and observers according to the needs of the experiment.
It can, among other things, 
* issue commands to actors
* request data from individual actors
* read data from an observer

Potentially a GUI could be attached here too.

## Action
An Action represents a method or function of the device represented by a Driver.
It can be called with zero or more arguments.

## Actor
The Actor is a component which (typically) instantiates one or more Driver class and/or does some kind of processing.
Actions which the actor can take could be anything from measuring on a regular basis ("acting" as a sensor), commanding hardware devices through the use of the driver to do X or Y, processing some data and re-emitting results (e.g. a PID controller).

> The actor is part of the communication network, that means it is listening for messages and maybe sending messages by itself.

Reviewer 2: I would put that part on top of the description, as I understand it as the main job:

The actor participates in the network listening for commands and acts on them. It may act regularly at its own account.
Typically an actor controls one or several devices via Drivers...

Author: Makes sense!

Potentially a GUI could be attached here too.

The Actor would send commands to a Driver, instructing it to `set`/`get`/`call` a given Parameter/Action by name, with a value if appropriate, and receive replies.

```{note}
We could achieve more complete separation and make the Actor purely about processing _inputs_, doing some _operation_ and generating _outputs_.
This could be either stateless (temperature conversion) or stateful (PID control).
Maybe a better name would be "Processor" then.

> This means that _all_ device-related stuff stays compartmentalized in the Driver, and an Actor would address that Driver over the network (to fetch the Parameter values it is interested in).

Reviewer 2: If we go through the network again, how do we call the recipient of the messages?

I understood that the actor is some software with two parts: a communication part and a "working" part, for example a driver. (although I'm unsure whether actor means the whole conglomerate or just the communication part)

Author: Not sure I understood correctly, let me try to illustrate
```mermaid
sequenceDiagram
    par
        Actor1->>Driver42: GET temp_K
        Driver42->>instr-on-COM1: "TEMP?\n"
        instr-on-COM1->>Driver42: "275.14\n"
        Driver42->>Actor1: 275.14
    and
        Actor1->>Driver11: GET temperature
        Note over Driver11: Did fetch a fresh value 23ms ago, using cached value
        Driver11->>Actor1: 300.1
    end
    Note over Actor1: computes average temperature, converts to degC
    Note over Actor1: stores/caches value maybe
    Actor1->>Observer1: "Average oven temperature" 14.47 degC
```
Is this clear?

Reviewer 2: Yes

Author: So, agreed on putting the Device interaction all into Driver as I propose?

Author: > communicate in the same way as Actors and Observers - I do not know how to better separate those in mermaid though. Possibly it will be enough to "put a caption" to it in the sense of describing the abstractness of the arrows.

I noticed that, too. I can try to improve that (dashed arrows are possible I think).

> Students who never do anything if they did not get an input first (i.e. "process them") generally do not stay long.

:laughing: 

Yeah, I noticed that the boundaries felt weird/unclear, that's why asked for feedback on that. I don't think one or the other option is more "correct".

The motivation to separate this more, and to have the Driver participate in the network, was to

* get better separation of "business logic" (actor) and "HW comms layer" (driver)
* Have the Driver be a common interface for _all_ different instrument libraries to our protocol. I felt we should not put that into the Actor, too. 

You say
> An Actor is then a Component which participates in the ECP communication (contrary to the Driver), which may _use_ a Driver instance for its purposes
 
That "use" is the key -- how do we define that? How can an Actor _use_ a driver? What are a driver's capabilities? We are trying to create a not-Python-specific thing here, so we should not define this in terms of classes, properties, etc.
If we make the Driver a component, and define its interaction with the rest of ECP via _messages_, we can do that, people can code up a conforming implementation in their favourite language/style, and we can interoperate.

This was a bit stream-of-consciousness, but does that make the motivation clearer? How can I improve the text toward that (if you agree)?

Reviewer 2: Let us continue in #17, to keep that discussion in one place, which can be found again.

Author: Resolved in #17 and with the updates, I think

This might contain or wrap a `pymeasure.Instrument` instance or something from another instrument library.
This is the place where all instrument libraries (including pymeasure) wire their hardware interface classes into ECP.

Concerning the ECP, we draw the abstraction boundary at the Driver -- the details on how this communicates with hardware (SCPI, dll, ...) should not be relevant for the protocol details.
What we define is how the other ECP components interact with the Driver, how it determines and announces its capabilities, etc.

A driver may contain/manage
* An object that communicates with the connected hardware device/instrument, including managing its lifecycle (init, operations, shutdown)
* The name of the connected instrument
* A list of available Parameters (properties/attributes to get/set) and Actions(methods to call)
* A list of Parameters to poll/publish regularly (and the interval for that)
* `set`/`get`/`get_all`/`call` interfaces (for incoming commands to use to act on Parameters and Actions)
* A cache of parameter values (to avoid unnecessary communication)

> * Maybe a task queue and/or concurrent access management/locking

Reviewer 2: It the actor/driver part is single threaded, (maybe using timers but the relevant work is in the main thread), we do not need queues, as the Zmq sockets serve as queues: It reads the next message after finishing the last one.

Author: Ah, of course, thanks!

* Logging configuration

```{note}
We might want to add the notion of "Channels", especially for the multi-Director stuff
```

# Parameter
A Parameter is a property (in the English, not the Pythonic sense) of the device represented by a Driver

> It has a name and can be read(`get`) or `set`, and recent values might be cached in the Driver.

Reviewer 2: Caching is a difficult topic: yes it is faster, but might not be true anymore.
I like pymeasure's stance to query the hardware, as that is the point of truth.
I see, however, that it might be good for some applications to cache values.

Author: I agree, that's why I said "might". In my tool I made good experience with it.
When the networks grow, we will have less control over the precise timing. If, let's say, 10 Parameters of an instrument are relevant/interesting, we might have Actors querying those every 500ms, a Observer for storage every 1000ms, a Director with a GUI every frame/16ms. That's a lot of requests and communication to re-fetch quite fresh data, which is what I want to avoid.

Another thing is that I had some instruments (registerbased) that could fetch the data for _all_ Parameters in _one_ request (very efficient!). I had a `get_all` method that made use of that for that instrument type, and updated all the cache values with that data, and the other participants (plot, PID, GUI) could then fetch the cached values really quickly.

We could foresee a configurable `cache_timeout` per Driver that controls how outdated data might become.

Reviewer 2: Regular readout and publishing values as they become available can reduce device connection calls.
Nevertheless cached values are good:
I just see the case of a triggered device (oscilloscope), where you might want to know the last results without waiting for the next trigger.

Reviewer 2: To be clear: I'm not against caching per se, but want to be careful.
And we have to weigh, whether the complexity and calls (calculate time difference...) are worth it.

Author: > I just see the case of a triggered device (oscilloscope), where you might want to know the last results without waiting for the next trigger.

my `get` method had a `from_cache=True` argument, so you could always force a fresh fetch if you need it.

Indeed, it needs weighing. I'll leave it in and we can refine later.

It may correspond closely to _attributes_ or Python (or PyMeasure) _properties_ of the instrument interface classes.

## Procedures
Sequences of commands that make up experiment runs, e.g. PyMeasure procedures.

> Procedures as they are now should maybe not rely on actors, but could, and again, need to look into it further). Once a user is familiar with this bigger framework, they can then add new actors, to grow their system one by one.

Reviewer 2: That sentence is broken

Author: Indeed :ouch:

```{note} This is a placeholder, we have not defined the concept yet
```

### Actor Driver separation
We want to keep Actor and Driver separated, to leave the entry threshold as low as possible, the learning curve flat, and the usability of ECP modular.
If someone wants to just connect with a Driver instance, directly, that should be possible, and when looking at the class, easily understood (and implemented).

> Once a user is familiar with that, and a slightly bigger system is envisaged by them, one or two Driver classes (which might already exist for this specific instrument in pymeasure, or have easy templates) might get packaged into Actors, and a director attached to them (via broker/proxy/whatever), so both can be controlled together (e.g. in a sequence-like fashion.

Reviewer 2: Small point to use different devices:
```suggestion
Once a user is familiar with that, and a slightly bigger system is envisaged by them, one or two Driver classes (which might already exist for these specific instruments in pymeasure, or have easy templates) might get packaged into Actors, and a director attached to them (via broker/proxy/whatever), so both can be controlled together (e.g. in a sequence-like fashion.
```

Author: Thanks (fixed locally)!

This page provides details on the main component/entity types that make up a deployment of the ECP.

## Director
A Director manages a part of or the entire setup, orchestrating the Components according to the needs of the experiment.
It can, among other things, 
* issue commands to Actors
* request data from individual Actors
* read data from an Observer

Potentially a GUI could be attached here too.

## Actor
A Actor is a component that interfaces with a (hardware) Device and that has a specific API on the ECP side.
A Actor must contain a Driver object dealing with communication with the Device.

We define how the other ECP components interact with the Actor, how it determines and announces its capabilities, etc.
An Actor implements the mapping between ECP messages and Driver calls/attribute access.

> A Actor must contain/manage/provide

Reviewer 2: ```suggestion
An Actor must contain/manage/provide
```

Next line should start with "A Driver"

Author: Ha, that's what you get for doing search/replace and not switching off the mental autocorrect 😅

* An Driver that communicates with the connected Device, including managing its lifecycle (init, operations, shutdown)
* The name of the connected instrument
* A list of available Parameters (properties/attributes to get/set) and Actions (methods to call)
* `set`/`get`/`get_all`/`call` interfaces (for incoming commands to use to act on Parameters and Actions)

> A Actor may contain/manage

Reviewer 2: ```suggestion
An Actor may contain/manage
```

This is a placeholder, we have not fleshed out the concept yet
:::

## Processor
The processor runs some kind of processing operation on one or more inputs and produces one or more outputs  
It can be stateless (e.g. temperature conversion) or stateful (like a PID controller).
It may act regularly of its own accord.

> The Processor would send commands to a Actor, instructing it to `set`/`get`/`call` a given Parameter/Action by name, with a value if appropriate, and receive replies.

Reviewer 2: ```suggestion
The Processor would send commands to an Actor, instructing it to `set`/`get`/`call` a given Parameter/Action by name, with a value if appropriate, and receive replies.
```

:::

## Coordinator
A component primarily concerned with routing/coordinating the message flow between other Components.
It represents the intermediate zmq brokers, proxies or somesuch.
Multiple coordinator instances may be necessary for large deployments, but a single coordinator instance should be sufficient for operation.

The presence of a coordinator should avoid the complexity/scaling of a purely point-to-point messaging approach. 

## Observer
A Component that receives data from other Components, e.g. for logging, storage, or plotting, either directly in a streaming fashion, batched, or delayed.
It only consumes message streams, but does not command `Actors`.

:::{note} Depending on setup, some commanding might be necessary, e.g. to subscribe/register.
:::

## Notes 
### Complexity scaling
We want to leave the entry threshold as low as possible, the learning curve flat, and the usability of ECP modular.
If someone wants to just connect with a Actor instance, that should be possible, and when looking at the class, easily understood, as the Actors have a consistent API.
This assumes that the library-specific Actor has already been written by the library maintainers.

> Once a user is familiar with that, and a slightly bigger system is envisaged by them, multiple Actor classes might be used, maybe with a Processor doing some useful transformations, and a director attached to them (via broker/proxy/whatever), so all can be controlled together.

Reviewer 2: ```suggestion
Once a user is familiar with that, and a slightly bigger system is envisaged by them, multiple Actor classes might be used, maybe with a Processor doing some useful transformations, and a Director attached to them (via Coordinators), so all can be controlled together.
```

Possibly Procedures are added for sequencing, too.

> Once a user is familiar with this bigger framework, they can then add new Processors, to grow their system one by one.

Reviewer 2: You already mentioned Processors in the sentence above.
